sdk/trace/batch_span_processor.go

```diff
@@ -167,53 +156,96 @@ func WithBlocking() BatchSpanProcessorOption {
 	}
 }
 
-// processQueue removes spans from the `queue` channel until there is
-// no more data.  It calls the exporter in batches of up to
-// MaxExportBatchSize until all the available data have been processed.
-func (bsp *BatchSpanProcessor) processQueue(batch *[]*export.SpanData) {
+// processQueue removes spans from the `queue` channel until processor
+// is shut down. It calls the exporter in batches of up to MaxExportBatchSize
+// waiting up to ScheduledDelayMillis to form a batch.
+func (bsp *BatchSpanProcessor) processQueue() {
+	timer := time.NewTimer(bsp.o.ScheduledDelayMillis)
+	defer timer.Stop()
+
+	batch := make([]*export.SpanData, 0, bsp.o.MaxExportBatchSize)
+
+	exportSpans := func() {
+		if !timer.Stop() {
+			<-timer.C
+		}
+		timer.Reset(bsp.o.ScheduledDelayMillis)
+
+		if len(batch) > 0 {
+			bsp.e.ExportSpans(context.Background(), batch)
+			batch = batch[:0]
+		}
+	}
+
+loop:
 	for {
-		// Read spans until either the buffer fills or the
-		// queue is empty.
-		for ok := true; ok && len(*batch) < bsp.o.MaxExportBatchSize; {
-			select {
-			case sd := <-bsp.queue:
-				if sd != nil && sd.SpanContext.IsSampled() {
-					*batch = append(*batch, sd)
-				}
-			default:
-				ok = false
+		select {
+		case <-bsp.stopCh:
+			break loop
+		case <-timer.C:
+			exportSpans()
+		case sd := <-bsp.queue:
+			batch = append(batch, sd)
+			if len(batch) == bsp.o.MaxExportBatchSize {
+				exportSpans()
 			}
 		}
+	}
 
-		if len(*batch) == 0 {
-			return
+	go func() {
+		bsp.enqueueWait.Wait()
+		close(bsp.queue)
+	}()
+
+	for {
+		if !timer.Stop() {
+			<-timer.C
 		}
+		const waitTimeout = 30 * time.Second
```

Comment:
I don't think it should be - it is only needed in case there is a panic and enqueueWait.Done is not called (defensive programming since we already have a timer). I can remove this and code will continue working fine. It can be replaced with something like `2*ScheduledDelay` if you prefer.